docs_src/design/ucr/Record-and-Replay.md

```diff
@@ -0,0 +1,88 @@
+- Record and replay can be used to reproduce a reported bug.
+- Record and export could serve as a simple means to backup data in the event of data loss.
+
+### Target Users
+- Device Manufacturer
+- Device Owner
+- Device User
+- Device Maintainer
+- Software Developer
+- Software Deployer
+- Software Integrator
+
+### Description
+Target users have the need to be able to replay recorded EdgeX Event/Readings for functional, performance or reproducible testing. This UCR describes a new capability that allows user to first record Event/Readings from real devices in real-time and be able to replay the Event/Readings as if it was in real-time. The static device profile and device definition files at time of capture will need to be available and loaded at the time the captured data is replayed. 
+
+### Existing solutions
+There are simulators for some devices (i.e. Modbus), but there isn't an general solution to reproduce real device data into EdgeX without the physical devices being present. These simulators also don't have a way to produce a specific set of results in a timeline as do physical devices.
+
+### Requirements
+#### Record
+
+Shall have capability to record device Event/Reading data as received from the Device Service(s). 
+
+- Recorded Event/Reading timestamps shall be sufficient to determine the captured rate.
+
+- Recorded Events shall be sufficient to determine the major version of EdgeX used, i.e. ApiVersion
+
+- Record must allow the duration of capture and/or the max number of Readings to capture to be specified
+
+- Record must capture all Device definitions for devices referenced in the captured Event/Readings
+
+- Record must capture all Device Profiles for devices referenced in the captured Event/Readings
+
+- Record must have the ability to record data only from specified target devices or device profiles.
+
+#### Export
+
+Shall have capability to export captured data for use at a later time or to send to other users
+
+- Exported data shall be in JSON format with option for it to be compressed (ZIP or TAR). 
+
+- Exported binary Event/Readings shall not be re-encoded in CBOR.
+
+#### Import
+
+Shall have capability to import data that was previously recorded and exported
+
+- Import must validate that the current EdgeX version is compatible with the APIVersion captured in the recorded Events. 
+
+- Import must add any captured Device definitions and Device Profiles to the system prior to play back.
+
+- Import shall have option to overwrite or use existing Device definitions and Device Profiles
+
+#### Replay
+
+Shall have capability to replay previously captured data
+
+- Replay must allow captured data to be replayed at captured, slower or faster speeds
+
+- Replay must allow for the repeat replay option with number of times to repeat the captured data.
```

Comment:
Perhaps something to discuss as part of the ADR: I'm not sure if this is possible if timestamps are included in the readings. Will this result in new events (with new timestamps) containing same old readings? Maybe the replay tool can automatically shift the timestamps on replays.

Reply:
No, I think that is a good requirement. i.e timestamps must be adjusted so they are the current time when published.

Reply:
Added requirement

Reply:
The convenient implementation would be time delta from the start of capture. So saying "timestamp" might imply something different and harder to work with on replay than (picking a name) "time marks".

Reply:
Each Event and each Reading have a timestamp that need to be changed when replayed to be current time that Event/Readings were replayed. The ADR will get into how the original timestamps will be used to order and replay at recorded speed, slower or faster.